### DISCUSSION TITLE
Does Strongswan support EC:prime256v1 keys and certificate

### DISCUSSION THREAD

**Discussion by satishkrishna687** [strongswan/strongswan#1668] [Q&A]
I have my EC:prime256v1 key inside my OP-TEE and I am able to load it 

`Fri, 2023-04-21, 20:10:52.464 13[LIB] builder L0 CRED_PRIVATE_KEY - ANY of plugin 'pkcs11'
Fri, 2023-04-21, 20:10:52.484 13[CFG] found key on PKCS#11 token 'op-tee':0
Fri, 2023-04-21, 20:10:52.548 13[CFG] loaded ECDSA private key from token `

**but it fails to parse the certifacte that is stored in the OP-TEE or even in the /etc/swanctl/x509/**

these are the command I used to generate all the CA keys and User key and certificates
I have only used openssl command to generate sign the certificates

`-----------generate CA on HOST-----------------------------------
 **openssl ecparam -genkey -name prime256v1 -out strongswanKey.pem**
 
 //to see the contents of the above Key use the below command
 openssl ec -in strongswanKey.pem -noout -text
 
 
**openssl req -x509 -new -key strongswanKey.pem -out strongswanCert.pem -days 3652 -subj '/C=CH/O=strongSwan/CN=strongSwan Root CA'**
//to verify use this command
openssl x509 -in strongswanCert.pem -noout -text

--------------------generate HOST keys (moonKey, moonCSR and moonCert) for Host machine----------------

//generate users ec:prime256v1 key
**openssl ecparam -genkey -name prime256v1 -out moonkey.pem**

//CSR
>getit csr.cnf
[req]
distinguished_name = req_distinguished_name
req_extensions = v3_req
prompt = no

[req_distinguished_name]
C = CH
O = strongswan
CN = moon.strongswan.org

[v3_req]
subjectAltName = DNS:moon.strongswan.org

//command to generate csr
**openssl req -new -config csr.cnf -key moonKey.pem -outform pem -out moonCSR.pem**

//sign the CSR
**openssl x509 -in moonCSR.pem -out moonCert.pem -CA strongswanCert.pem -CAkey strongswanKey.pem -set_serial 01 -days 1826 -extfile <(printf "subjectAltName=DNS:moon.strongswan.org\n")**

-----------------------------------------Now OPTEE-------------------------------------------------

pkcs11-tool --module /usr/lib/libckteec.so.0 --list-slots

pkcs11-tool --module /usr/lib/libckteec.so.0 --init-token --label mytoken --so-pin 4321 --init-pin --pin 1234

pkcs11-tool --module /usr/lib/libckteec.so.0 --token-label mytoken --login --pin 1234 --keypairgen --label mykey --id 4114 --key-type EC:prime256v1

---be in the shared folder where we generated the aobve host and CA keys and certificates----------

>vi csr1.cnf
[req]
distinguished_name = req_distinguished_name
req_extensions = v3_req
prompt = no

[req_distinguished_name]
C = CH
O = strongswan
CN = sun.strongswan.org

[v3_req]
subjectAltName = DNS:sun.strongswan.org

----save the above file and follow the below commands---------

openssl
   openssl>**engine dynamic -pre SO_PATH:/usr/lib/engines-1.1/pkcs11.so -pre ID:pkcs11 -pre LIST_ADD:1 -pre LOAD -pre MODULE_PATH:/usr/lib/libckteec.so.0**

    **oenssl>req -new -config csr1.cnf -engine pkcs11 -key "pkcs11:object=mykey;type=private?pin-value=1234" -keyform engine -outform pem -out sunCSR.pem**

---now CSR is generated for the key stored inside OP-TEE as sunCSR.pem---

//sign the sunCSR with strongswanKey and strongswanCert
**openssl x509 -req -in sunCSR.pem -out sunCert.pem -CA strongswanCert.pem -CAkey strongswanKey.pem -set_serial 01 -days 1826 -extfile <(printf "subjectAltName=DNS:sun.strongswan.org\n")**

///Now sunCert is generated

convert it to der format to store inside OP-TEE

openssl x509 -in "sunCert.pem" -out "sunCert_copy.der" -outform der 

//now this will push the certificate inside op-tee
pkcs11-tool --module /usr/lib/libckteec.so.0 --token-label mytoken --login --pin 1234 --write-object "sunCert_copy.der" --type cert --id 4114 --label mycert``
===========================================================================================

it also fails even I store it locally
# swanctl --load-all
loading '/etc/swanctl/x509/sunCert_copy.der' failed: parsing X509 certificate failed
loading '/etc/swanctl/x509/sunCert.pem' failed: parsing X509 certificate failed
loaded key token-slot0 from token [keyid: 7c7d3d0b1e29b3a3c0594d3ca27def0ae5bd14f8]
no authorities found, 0 unloaded
no pools found, 0 unloaded
loaded connection 'host-psk'
successfully loaded 1 connections, 0 unloaded

Now I converted my sunKey to der format and stored it, it showed the same error

`# pkcs11-tool --module /usr/lib/libckteec.so.0 --login -O
Using slot 0 with a present token (0x0)
Logging in to "mytoken".
Please enter User PIN: 
Private Key Object; EC
  label:      mykey
  ID:         4114
  Usage:      sign, derive
  Access:     sensitive, always sensitive, never extractable, local
Public Key Object; EC  EC_POINT 256 bits
  EC_POINT:   044104d7824d6f8ce569dd457f6c882b88b2e60e178a28503417dee913b77b3c24954d7cd903f6b0ea535730c481c3f5052bdcbc84c97295d7926003ef6bbe3e591c15
  EC_PARAMS:  06082a8648ce3d030107
  label:      mykey
  ID:         4114
  Usage:      verify, derive
  Access:     local
# pkcs11-tool --module /usr/lib/libckteec.so.0 --token-label mytoken --login --p
in 1234 --write-object sunCert_copy.der --type cert --id 4114 --label mykey
Created certificate:
Certificate Object; type = X.509 cert
  label:      mykey
  subject:    DN: C=CH, O=strongswan, CN=sun.strongswan.org
  ID:         4114
# vi /etc/swanctl/swanctl.conf 
# ipsec restart
Stopping strongSwan IPsec...
Starting strongSwan 5.9.8 IPsec [starter]...

# swanctl --load-all
**loading '/etc/swanctl/x509/sunCert_copy.der' failed: parsing X509 certificate failed
loading '/etc/swanctl/x509/sunCert.pem' failed: parsing X509 certificate failed
loaded key token-slot0 from token [keyid: 7c7d3d0b1e29b3a3c0594d3ca27def0ae5bd14f8]**
no authorities found, 0 unloaded
no pools found, 0 unloaded
**loading connection 'host-psk' failed: unable to load certificate: cert
loaded 0 of 1 connections, 1 failed to load, 0 unloaded**
`\
LOG:
Fri, 2023-04-21, 20:40:39.386 15[LIB] builder L1 CRED_CERTIFICATE - X509 of plugin 'pkcs11'                    
Fri, 2023-04-21, 20:40:39.386 15[LIB] builder L1 CRED_CERTIFICATE - X509 of plugin 'x509'                                            
**Fri, 2023-04-21, 20:40:39.387 15[CFG] parsing PKCS#11 certificate 41:14 failed**   
=====================================================================

My question is since it loads my private key  , why is not supporting my certidficate

Details of my certificate satish@satish-vm:~/optee-qemu/all_keys$ openssl x509 -in sunDer.der -noout -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 1 (0x1)
        Signature Algorithm: ecdsa-with-SHA256
        Issuer: C = CH, O = strongSwan, CN = strongSwan Root CA
        Validity
            Not Before: Apr 21 18:06:12 2023 GMT
            Not After : Apr 20 18:06:12 2028 GMT
        Subject: C = CH, O = strongswan, CN = sun.strongswan.org
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:d7:82:4d:6f:8c:e5:69:dd:45:7f:6c:88:2b:88:
                    b2:e6:0e:17:8a:28:50:34:17:de:e9:13:b7:7b:3c:
                    24:95:4d:7c:d9:03:f6:b0:ea:53:57:30:c4:81:c3:
                    f5:05:2b:dc:bc:84:c9:72:95:d7:92:60:03:ef:6b:
                    be:3e:59:1c:15
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        X509v3 extensions:
            X509v3 Subject Alternative Name: 
                DNS:sun.strongswan.org
            X509v3 Subject Key Identifier: 
                7C:7D:3D:0B:1E:29:B3:A3:C0:59:4D:3C:A2:7D:EF:0A:E5:BD:14:F8
            X509v3 Authority Key Identifier: 
                A8:58:D9:33:FC:A2:64:74:63:54:AE:74:32:86:73:67:F7:C2:F7:90
    Signature Algorithm: ecdsa-with-SHA256
    Signature Value:
        30:46:02:21:00:a7:0e:01:82:63:5f:a2:5d:a1:9a:78:08:d7:
        1a:15:d7:b9:57:0e:8b:c7:b3:bf:ae:73:37:ec:f7:7f:9c:d1:
        18:02:21:00:80:ff:d2:8d:7a:b5:9e:1e:4b:d2:22:99:88:1b:
        8f:13:40:eb:1f:c5:61:85:fe:1b:84:cd:5a:a6:4c:a7:85:45

**Comment by tobiasbrunner** [comment]
ECDSA keys/certificates are generally only supported if the _openssl_ plugin is loaded in strongSwan (alternatives are the _botan_ or the _wolfssl_ plugins). Depending on the PKCS#11 module it might also be possible to use that as provider for public key operations (requires enabling `use_pubkey` and `use_ecc` in the config for the [_pkcs11_ plugin](https://docs.strongswan.org/docs/5.9/plugins/pkcs11.html)). Note that the latter could be quite slow.
